### ISSUE TITLE
Proposal: CI check to flag module-level names dropped by a refactor

### ISSUE THREAD

**Issue by shlokkvaishnav** [kornia/kornia#4022]
Follow-up to @ducha-aiki's comment on #4003 (closed) — the real problem behind #3986 wasn't the missing `pad` symbol itself, it's that nothing catches a refactor silently removing a name a module used to expose (whether or not it was ever in `__all__`).

## Proposal

A CI check, run on every PR, that:

1. For each changed file under `kornia/**/*.py`, statically parses (via `ast`, no import/execution needed) the set of module-level names bound in the file — top-level `def`/`class`, assignments, and `import`/`from ... import` targets — at both the PR branch tip and the merge-base commit.
2. Diffs the two sets per file and reports any name present at the merge-base but missing at the tip, split into two severities:
   - **Removed from `__all__`**: a guaranteed break of documented public API — hard fail.
   - **Removed from module scope entirely, but was never in `__all__`** (the `pad` case): informational only — surfaced as a PR comment/annotation, not a blocking failure, since plenty of these removals are intentional internal cleanup.
3. No new runtime dependency: this only needs `ast` + `git show <merge-base>:<path>` from the stdlib, and only runs in CI (never imported into `kornia/`), so it doesn't touch the "only PyTorch as a third-party dependency" rule for the library itself.

This would have caught #3986 automatically as an informational note on the PR that dropped `pad` from `kornia/core`, and would hard-fail on the more serious case of an actual `__all__` regression, without requiring anyone to remember to check by hand.

Open questions I'd want scope agreement on before implementing:
- Should the informational (never-`__all__`) case block the PR (with an explicit override label/comment) or just annotate it? I'd lean toward annotate-only to start, given how often that kind of "accidental" surface is legitimately fine to drop.
- Scope to `kornia/` only, or also `testing/`?
- Where should this live — a script under `scripts/`, or a dedicated GitHub Action?

Happy to open a PR once the shape here is agreed on.

**Comment by shlokkvaishnav** [comment]
Drafted an implementation of this to make the shape concrete: https://github.com/shlokkvaishnav/kornia/tree/feat/module-surface-diff-check

- `scripts/check_module_surface_diff.py` -- pure `ast` + `git show`, no import of kornia needed
- `tests/scripts/test_check_module_surface_diff.py` -- unit tests including an end-to-end test against a throwaway git repo reproducing the exact `pad`/#3986 shape (undocumented name dropped) alongside a documented `__all__` removal
- `.github/workflows/module-surface-diff.yml` -- runs on PRs touching `kornia/**/*.py`

Currently: `__all__` removals hard-fail, undocumented-name removals are `::notice`-only. Holding off on opening a PR until there's agreement on the open questions above (annotate vs. block, `testing/` scope, script vs. dedicated Action) -- happy to adjust based on what you'd prefer.

**Comment by ducha-aiki** [comment]
Yes — please build this. The shape is agreed, and answers to your three questions are below, along with review notes from actually running your branch.

First, an apology. While this issue was open and waiting for scope agreement, we opened #4027 doing the same thing — same design, same `ast` + `git show` approach, same `__all__`-fails / undeclared-informational split. Yours was filed at 11:24 UTC, ours at ~11:38. You did exactly what `CONTRIBUTING.md` asks — proposed first, asked before implementing, drafted to make it concrete — and got landed on from the maintainer side for it. #4027 is closed. This is yours.

## Your three open questions

**1. Annotate or block for the never-`__all__` case?** Annotate only, as you leaned. Your instinct is right and the numbers back it: an inventory of the tree finds **2143** names that are importable purely because a module does `from torch import ...` or similar, plus **1148** more in the 368 of 500 modules that declare no `__all__` at all. Hard-failing on those makes every incidental import permanent API, and the check gets switched off within a month. A notice a reviewer reads beats a gate a maintainer disables. Note the honest consequence, which is worth putting in the PR description: `pad` itself would have produced a notice, not a red build. That is still the outcome that would have helped — the reviewer of that PR would have seen the line.

**2. `kornia/` only, or `testing/` too?** `kornia/` only. `testing/` is support code for our own suite with no external contract.

**3. Script or dedicated Action?** A script under **`.github/scripts/`**, next to `check_doc_links.py`, with a workflow modelled on `.github/workflows/doc-links.yml`. That is the established precedent for exactly this kind of stdlib-only checker. Not a new top-level `scripts/`, and not a dedicated Action — an Action is packaging overhead for something that is one `python3` invocation.

## Review notes from running your branch

I checked out `feat/module-surface-diff-check` and ran it against real revision pairs. It works, and it correctly flags `upscale_double` — a documented function deleted by a performance PR, which is the most serious thing the audit turned up. Three fixes:

**a. Exempt modules with a module-level `__getattr__`.** This is the one real false positive. Run against `main` with `--base-ref v0.8.2`:

```
::error file=kornia/utils/__init__.py::Removed from __all__: ['_extract_device_dtype', 'deprecated']
```

But `kornia.utils.deprecated` resolves fine at runtime — `kornia/utils/__init__.py` has a module-level `__getattr__` deprecation shim (added in #3898) that serves it with a `DeprecationWarning`. That shim is the *correct* pattern, the one thing in the 0.8.3 churn that was done right, so the check must not punish it. Suggest: if the new revision's module defines `__getattr__` at module scope, downgrade or skip its removals.

**b. Recurse into `Try` and `If` bodies — with one exception.** `parse_module_surface` iterates `tree.body` only, so names bound inside `try: ... except ImportError:` or a version-conditional `if` are never collected, and a removal there goes unreported. Worth recursing. The exception: **`if TYPE_CHECKING:` bodies must stay excluded** — those names are not importable at runtime, so they are neither surface nor regression. Your current code excludes them by accident (it doesn't descend into `If` at all); once you add recursion, make that exclusion deliberate.

**c. Don't hard-fail on `kornia.contrib`.** `docs/source/get-started/stability.rst` puts it in the Experimental tier — "No stability promise." Report those removals, never block on them. Without this, the `v0.8.2 → v0.8.3` comparison produces 396 contrib rows.

Two things you already got right and should keep: the three-dot `git diff base...HEAD` (that is merge-base semantics, which is what you want — it sees what the PR removes, not what the base branch already lost), and the `_changed_kornia_files` comment about git's `**` pathspec. The workflow will need `fetch-depth: 0` on checkout for the merge base to be reachable.

## One addition worth making, with the payoff already proven

Add the inverse check: a name **in `__all__` with no module-scope binding**. It costs a few lines on top of what you have, and it is not hypothetical — three exist on `main` right now:

```
kornia.color.AUTUMN · kornia.core.eye_like · kornia.core.vec_like
```

Each makes `from <module> import *` raise `AttributeError` today, and `tests/test_api_surface.py` passes anyway because it reads `__all__` without checking that the names resolve.

That is filed as **#4023**, and it is yours if you want it — it is a good small one: `kornia/core/__init__.py` needs `from .ops import eye_like, vec_like`, `AUTUMN` is a stale `__all__` entry to delete, and `tests/test_api_surface.py` needs the resolvability assertion. It can be a separate PR before or after this one, whichever you prefer.

## What not to spend time on

So you don't sink effort into things that are closed or held:

- **#4024** (`kornia.core`'s 32 removed names) — closed, not resurrecting. That module was an artifact of an abandoned attempt to run kornia on multiple backends (PyTorch and TensorFlow); once that was dropped, the names were pass-throughs to torch with no purpose.
- **#4025** (`upscale_double`) — closed, stays removed. It was a leftover from the OpenCV-mimicking DoG implementation, superseded when that code was improved.
- **#4026** — groups 1, 3 and 4 are maintainer-held pending the repair-and-deprecation window. Group 2 there (the accidental `logger`/`np`/`ort`/`url` names) is available and mechanical, if you want another small one.

Thanks for pushing past the immediate symptom in #4003 to the thing that actually caused it. That is the more useful contribution, and it is the one that will land.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by shlokkvaishnav** [comment]
Opened as #4029, with the three fixes applied and validated against the exact regressions and false positive from your review.
